bin/exec.ml

```diff
@@ -49,105 +49,80 @@ let term =
-            Some prog
-          else if not Sys.win32 then
-            None
+      let build_prog p =
+        let open Memo.Build.O in
+        if no_rebuild then
+          if Path.exists p then
+            Memo.Build.return p
           else
-            let prog = Path.extend_basename prog ~suffix:Bin.exe in
-            Option.some_if (Path.exists prog) prog
+            User_error.raise
+              [ Pp.textf
+                  "Program %S isn't built yet. You need to build it first or \
+                   remove the --no-build option."
+                  prog
+              ]
+        else
+          let+ () = Build_system.build (Action_builder.path p) in
+          p
       in
-      (* Good candidates for the "./x.exe" instead of "x.exe" error are
-         executables present in the current directory *)
-      let hints () =
-        let+ candidates =
-          let path = path_relative_to_build_root "" in
-          let+ targets =
-            Build_system.run (fun () -> Build_system.targets_of ~dir:path)
+      let not_found () =
+        let open Memo.Build.O in
+        let+ hints =
+          (* Good candidates for the "./x.exe" instead of "x.exe" error are
+             executables present in the current directory *)
+          let+ candidates =
+            Build_system.targets_of ~dir:(Path.build dir)
+            >>| Path.Set.to_list
+            >>| List.filter ~f:(fun p -> Path.extension p = ".exe")
+            >>| List.map ~f:(fun p -> "./" ^ Path.basename p)
           in
-          Path.Set.to_list targets
-          |> List.filter ~f:(fun p -> Path.extension p = ".exe")
-          |> List.map ~f:(fun p -> "./" ^ Path.basename p)
+          User_message.did_you_mean prog ~candidates
         in
-        User_message.did_you_mean prog ~candidates
-      in
-      match (real_prog, no_rebuild) with
-      | None, true -> (
-        Memo.Build.run (Memo.Lazy.force targets) >>= function
-        | [] ->
-          let* hints = hints () in
-          User_error.raise ~hints [ Pp.textf "Program %S not found!" prog ]
-        | _ :: _ ->
-          User_error.raise
-            [ Pp.textf
-                "Program %S isn't built yet. You need to build it first or \
-                 remove the --no-build option."
-                prog
-            ])
-      | None, false ->
-        let* hints = hints () in
         User_error.raise ~hints [ Pp.textf "Program %S not found!" prog ]
-      | Some real_prog, _ ->
-        let real_prog = Path.to_string real_prog in
-        let argv = prog :: args in
-        restore_cwd_and_execve common real_prog argv
-          (Super_context.context_env sctx))
+      in
+      let* prog =
+        let open Memo.Build.O in
+        Build_system.run (fun () ->
+            match Filename.analyze_program_name prog with
+            | In_path -> (
+              Super_context.resolve_program sctx ~dir ~loc:None prog
+              >>= function
+              | Error _ -> not_found ()
```

Comment:
At the first sight it seems bad to drop the error on the floor, because it might contain something useful for the user. On further inspection, it turns out that `Not_found` is the only possible error, so adding a type annotation here would make it more obvious that we don't ignore other kinds of errors.

Reply:
Sounds good, I did that